proposals/NNNN-retroactive-protocol-refinement.md

```diff
@@ -0,0 +1,100 @@
+
+```Swift
+extension FixedWidthInteger: ExpressibleByUnicodeScalarLiteral {
+  @_transparent
+  public init(unicodeScalarLiteral value: Unicode.Scalar) {
+    self = Self(value.value)
+  }
+}
+```
+In practical terms, for the Swift compiler, this effectively means creating ad-hoc extensions on Int8, Int16 and all types which conform to FixedWidthInteger along with their associated witnesses. It's more powerful however as an idea and in more abstract terms (from the Generics Manifesto):
+
+```
+protocol P {
+  func foo()
+}
+
+protocol Q {
+  func bar()
+}
+
+extension Q : P { // Make every type that conforms to Q also conforms to P
+  func foo() {    // Implement `P.foo` requirement in terms of `Q.bar`
+    bar()
+  }
+}
+
+func f<T: P>(t: T) { ... }
+
+struct X : Q {
+  func bar() { ... }
+}
+
+f(X()) // okay: X conforms to P through the conformance of Q to P
+```
+```
+This is an extremely powerful feature: it allows one
+to map the abstractions of one domain into another
+domain (e.g., every Matrix is a Graph).
+```
+It continues though:
+
+```
+However, similar to private conformances, it puts a
+major burden on the dynamic-casting runtime to chase
+down arbitrarily long and potentially cyclic chains
+of conformances, which makes efficient implementation
+nearly impossible.
+```
+The PR above strives to demonstrate that it may be possible to avoid these pitfalls and refine protocols retroactively at compile time, providing summary conformance data to the existing runtime for an efficient implementation of dynamic casts.
+
+## Proposed solution
+
+Allow conformances to be added to protocol extensions with all that implies in terms of all nominals conforming to the protocol being extended acquiring the added conformances. It would be useful if this also worked for synthesised conformances (and indeed it does).
+
+## Detailed design
+
+The PR above makes changes in three principal areas internal to the compiler.
+
+1)  The ConformanceLookupTable — a cache of the protocols a nominal conforms to has been adapted to traverse the inherited conformances of protocol extensions.
+
+2) The "Generic Signature" builder and module deserialising code has been adapted to traverse these extended sources of conformances.
+
+3) While the above takes place, a registry of extended conformances is kept in the compiler. This is used to emit the ad-hoc witness tables used by extended conformances in a manner that is compatible across modules.
+
+In terms of what is surfaced to the language, extending conformances of a protocol is fully functional including across modules.
+
+## Source compatibility
+
+This is an additive feature that has a syntax which while it follows of the norms of extensions in the Swift language it is currently not permitted. Therefore, there is no effect on source compatibility.
```

Comment:
```suggestion
This is an additive feature that has a syntax which while it follows the norms of extensions in the Swift language it is currently not permitted. Therefore, there is no effect on source compatibility.
```

Reply:
Thanks for taking a look @yg . I was tempted to hold out on some of the quaint english spellings but went with the flow.